### PR TITLE
ingestion of test/sandbox courses

### PR DESCRIPTION
**Pull request by shanbady**
### What are the relevant tickets?

Closes https://github.com/mitodl/hq/issues/7170

### Description (What does it do?)
This PR gives us the ability to mark specific courses as "test mode" courses which means - they are un-published (dont display anywhere) but still have their contentfiles pulled down and ingested into opensearch and qdrant (and therefore exposed via apis)

### How can this be tested?
1. checkout this branch
2. find some course that you know has contentfiles and delete the contentfiles associated with it. something like:
```python
from learning_resources.models import ContentFile
import random
run = ContentFile.objects.all()[random.randrange(ContentFile.objects.count())].run
learning_resource = run.learning_resource
print("id - ", learning_resource.id)
print("etl source - ",learning_resource.etl_source)
run.content_files.all().delete()
```
3. make note of the etl source and id for the resource (the script above prints it) and run that ETL's content file backpopulate management command with the --test-resource-ids flag. For example: 
```bash
python manage.py backpopulate_xpro_files --test-ids 123
```
4. This should also work for commands that allow for loading contentfiles for one off courses:
```
 python manage.py  backpopulate_mitxonline_files --resource-ids 123 --test-ids 123
```
5. Once that command finishes running, note that the resource will now have published=False and test_mode=True set on it:
```python
from learning_resources.models import ContentFile, LearningResource
lr = LearningResource.objects.get(id=123)
print(lr.test_mode, lr.published)
```
6. Visit the contentfiles endpoint and note that contentfiles have been pulled down for that resource. http://open.odl.local:8063/api/v1/contentfiles/?resource_id=**<resource_id>**
7. run the generate embeddings command and note that the contentfile shows up in the qdrant collection via the [dashboard](http://localhost:6333/dashboard#/collections/resource_embeddings.content_files):
```bash
python manage.py generate_embeddings  --resource-ids 123
```

# Additional Context:
1. There is mixin class that adds the --test-ids flag to all the relevant content file backpopulate commands. The mixin also sets the test_mode resource to published=False before pulling down content.

### Checklist:
- [ ] test for any undesirable edge cases in situations where the  learning resource is unpublished but run and associated content files are indexed.
- [ ] Does this PR cover all the places where we would need to check (learning_resource.published | learning_resource.test_mode)?

### CODE REVIEW

File: learning_resources/management/commands/mixins.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+            "--test-ids",
+            dest="test_ids",
+            help="List of readable IDs to use for testing",
+        )
+
+    def configure_test_resources(self, options):
+        if options.get("test_ids"):
+            test_ids = options["test_ids"].split(",")
+            LearningResource.objects.filter(id__in=test_ids).update(
+                test_mode=True, published=False
```
Comment (mbertrand) [inline learning_resources/management/commands/mixins.py:16]: During the standard scheduled ETL pipeline runs, might be good to set `test_mode=False` if `published` gets set to `True` (indicating that the course is no longer in "test" mode and should no longer be treated differently than other courses).

File: learning_resources/management/commands/mixins.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        parser.add_argument(
+            "--test-ids",
+            dest="test_ids",
+            help="List of readable IDs to use for testing",
+        )
+
+    def configure_test_resources(self, options):
+        if options.get("test_ids"):
+            test_ids = options["test_ids"].split(",")
+            LearningResource.objects.filter(id__in=test_ids).update(
```
[the remark this answers is not in the document]
Comment (shanbady) [reply learning_resources/management/commands/mixins.py:15]: regarding the runs filtering in edx_shared. would it make sense to do something like the following?

```python
runs = LearningResourceRun.objects.filter( 
     learning_resource__etl_source=etl_source, 
     learning_resource_id__in=ids
).filter(Q(published=True) | Q(learning_resource__test_mode=True))
```

File: learning_resources/management/commands/mixins.py (hunk 1 above)
Comment (shanbady) [reply learning_resources/management/commands/mixins.py:17]: I've made it so that all the etl management commands (both file and data) do this at the end 
```python
LearningResource.objects.filter(published=True, test_mode=True).update(test_mode=False)
```

File: learning_resources/management/commands/mixins.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbertrand) [reply learning_resources/management/commands/mixins.py:16]: That looks good to me

File: learning_resources/etl/mitxonline.py [hunk 3]
```diff
@@ -139,9 +138,7 @@ def extract_courses():
             _fetch_data(
                 settings.MITX_ONLINE_COURSES_API_URL,
                 params={
-                    "page__live": True,
```
Comment (mbertrand) [inline learning_resources/etl/mitxonline.py:142]: When running `backpopulate_mitxonline_data` on a fresh database, I'm getting this error:

```bash
celery-1                     |     "published": bool(course_run["is_enrollable"] and course["page"]["live"]),
celery-1                     |                                                       ~~~~~~~~~~~~~~^^^^^^^^
celery-1                     | TypeError: 'NoneType' object is not subscriptable
```

Seems like sometimes `course['page']` is `None`  now that more results are coming through from the API.  Try changing that line to:

```python
"published": bool(course_run.get("is_enrollable", False) and (course.get("page") or {}).get("live", False)),
```


After that a new exception occurs:

```bash
celery-1                     |   File "/src/learning_resources/etl/mitxonline.py", line 360, in transform_programs
celery-1                     |     for course in _fetch_courses_by_ids(program["courses"])
celery-1                     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery-1                     |   File "/src/learning_resources/etl/mitxonline.py", line 329, in _fetch_courses_by_ids
celery-1                     |     return list(
celery-1                     |            ^^^^^
celery-1                     |   File "/src/learning_resources/etl/mitxonline.py", line 47, in _fetch_data
celery-1                     |     ).json()
celery-1                     |       ^^^^^^
celery-1                     |   File "/opt/venv/lib/python3.12/site-packages/requests/models.py", line 978, in json
celery-1                     |     raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
```

**Review by mbertrand — CHANGES_REQUESTED** [review]

**Review by mbertrand — APPROVED** [review]
👍
